### PR TITLE
Feature/36 sicp module

### LINKED ISSUE #36: SICP algorithm development
Implement SICP as per prototype in ticket #35

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** bsk-000
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
SICP algorithm implementation, test, and documentation 

## Verification
Unit test containing several metrics

## Documentation
rst documentation and reference paper attached

## Future work
Nearest Neighbor algorithm ticket added to assess future improvements

### CODE REVIEW

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+/*! Compute the closest reference point to each measured point
+ @return void
+ @param Eigen::MatrixXd R_kmin1 : The current rotation to apply to the measured cloud
+ @param Eigen::MatrixXd t_kmin1 : The current translation to apply to the measured cloud
+ @param Eigen::MatrixXd s_kmin1 : The current scale to apply to the measured cloud
+ @param Eigen::MatrixXd measuredPoints : The measured point cloud
+ @param Eigen::MatrixXd referencePoints : The reference point cloud
+ */
+void ScalingIterativeClosestPoint::computePointCorrespondance(const Eigen::MatrixXd R_kmin1,
+                                                              const Eigen::MatrixXd t_kmin1, const double s_kmin1,
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: :large_orange_diamond: Parameter can drop to a new line.

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+                                                              const Eigen::MatrixXd measuredPoints,
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: :small_red_triangle_down: Should we make all these Eigen params, const reference params? e.g. `const Eigen::MatrixXd& measuredPoints`, this way the variable isn't copied to the function at each invocation. Should be faster.

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+}
+
+/*! Center point cloud around the average point
+ @return void
+ @param Eigen::MatrixXd measuredPoints : The measured point cloud
+ @param Eigen::MatrixXd referencePoints : The reference point cloud
+ */
+void ScalingIterativeClosestPoint::centerCloud(const Eigen::MatrixXd measuredPoints){
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: 🔻 const reference param here as well?

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    tkSum2 /= this->Np;
+    return tkSum1 - tkSum2;
+}
+
+
+/*! This module reads the data and checks for validity. If new information is present it
+ @return void
+ @param CurrentSimNanos The clock time at which the function was called (nanoseconds)
+ */
+void ScalingIterativeClosestPoint::UpdateState(uint64_t CurrentSimNanos)
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp:191]: :large_orange_diamond: What are your thoughts on making all the functions which support the UpdateState computation `private`?

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    void Reset(uint64_t CurrentSimNanos);
+    void computePointCorrespondance(const Eigen::MatrixXd R_kmin1, const Eigen::MatrixXd t_kmin1, const double s_kmin1,
+                                    const Eigen::MatrixXd measuredPoints, const Eigen::MatrixXd referencePoints);
+    void centerCloud(const Eigen::MatrixXd measuredPoints);
+    Eigen::MatrixXd computeRk(const double s_kmin1, const Eigen::MatrixXd R_kmin1);
+    double computeSk(const Eigen::MatrixXd R_kmin1);
+    Eigen::MatrixXd computeTk(const double s_k, const Eigen::MatrixXd R_k, const Eigen::MatrixXd measuredPoints);
+
+
+        Message<PointCloudMsgPayload> outputPointCloud;  //!< The output fitted point cloud
```
Comment (patkenneally) [inline src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.h]: :large_orange_diamond: This little guy is trying to escape off to the right

**Review by patkenneally — APPROVED** [review]

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp (hunk 1 above)
Comment (thibaudteil) [reply src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: Done

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp (hunk 2 above)
Comment (thibaudteil) [reply src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: Done

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp (hunk 4 above)
Comment (thibaudteil) [reply src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp:191]: Good idea and done

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.h (hunk 5 above)
Comment (thibaudteil) [reply src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.h]: Back in line!

File: src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp (hunk 3 above)
Comment (thibaudteil) [reply src/fswAlgorithms/pointCloudProcessing/SICP/scalingIterativeClosestPoint.cpp]: Done

**Pull request merged**